src/main/java/com/coreos/jetcd/EtcdWatchImpl.java

```diff
@@ -0,0 +1,294 @@
+     * @param watchOption key option
+     * @param callback    call back
+     * @return ListenableFuture watcher
+     */
+    @Override
+    public synchronized ListenableSetFuture<Watcher> watch(ByteString key, WatchOption watchOption, Watcher.WatchCallback callback) {
+        WatchRequest request = optionToWatchCreateRequest(key, watchOption);
+        Watcher.Builder builder = Watcher.newBuilder().withCallBack(callback)
+                .withKey(key)
+                .withWatchOption(watchOption);
+        ListenableSetFuture<Watcher> waitFuture = new ListenableSetFuture<>(null);
+        this.pendingWatchers.add(new Pair<>(builder, waitFuture));
+        getRequestStream().onNext(request);
+        return waitFuture;
+    }
+
+    /**
+     * Cancel the watch task with the watcher, the onCanceled will be called after successfully canceled.
+     *
+     * @param watcher the watcher to be canceled
+     */
+    @Override
+    public void cancelWatch(Watcher watcher) {
+        Watcher temp = watchers.get(watcher.getWatchID());
+        if (temp != null) {
+            synchronized (temp) {
+                if (this.watchers.containsKey(temp.getWatchID())) {
+                    this.watchers.remove(temp.getWatchID());
+                    this.cancelWatchers.put(temp.getWatchID(), temp);
+                    WatchCancelRequest cancelRequest = WatchCancelRequest.newBuilder().setWatchId(watcher.getWatchID()).build();
+                    WatchRequest request = WatchRequest.newBuilder().setCancelRequest(cancelRequest).build();
+                    this.requestStream.onNext(request);
+                }
+            }
+        }
+    }
+
+    /**
+     * empty the old request stream, watchers and resume the old watchers
+     * empty the cancelWatchers as there is no need to cancel, the old request stream has been dead
+     */
+    private void resume() {
+        synchronized (this) {
+            this.requestStream = null;
+            Watcher[] resumeWatchers = (Watcher[]) watchers.values().toArray();
+            this.watchers.clear();
+            this.cancelWatchers.clear();
+            resumeWatchers(resumeWatchers);
+        }
+
+    }
+
+    /**
+     * single instance method to get request stream, empty the old requestStream, so we will get a new
+     * requestStream automatically
+     * <p>the responseStream will distribute the create, cancel, normal response to
+     * processCreate, processCanceled and processEvents
+     * <p>if error happened, the requestStream will be closed by server side, so we call resume to resume
+     * all ongoing watchers
+     *
+     * @return
+     */
+    private StreamObserver<WatchRequest> getRequestStream() {
+        if (this.requestStream == null) {
+            synchronized (this) {
+                if (this.requestStream == null) {
+                    StreamObserver<WatchResponse> watchResponseStreamObserver = new StreamObserver<WatchResponse>() {
```

Comment:
I'm not sure if this object hold any status, so I new one for new requestStream.